### ISSUE TITLE
Secure device certificate and key

### ISSUE THREAD

**Issue by amit12cool** [Azure/iotedge#7062]
I have a software solution for IoT edge i.e. I have a exe and rpm installer that consists of the edge installation. The exe and the rpm will be installed on windows and Linux machine respectively. 

I need to secure my edge device cert and private key on the windows and Linux machine, so that if compromised one level of security is at least rather than not having any security.

Possible solutions I was thinking of:-

- Fetch a .pfx file from PKI and use that in edge config.toml. I think edge doesn't work with .pfx file and it needs the cert and private key to connect to IoT Hub. Also I didn't see any option in config.toml to mention the passphrase of the pfx file so that edge can decrpyt the pfx file itesef. Please confirm on this understanding is correct or not.

- I was thinking to have my custom encryption algorithm to encrypt the cert and private key and decrypt that during the installation process of edge before doing the provisioning. Once provisioning is done by edge device, I will delete the decrypted files. Question on this is - _Is it that edge requires the certificate again anytime, as I think the provisioning is done only once and not again and again and cert and private key is meant for provisioning only._

- I was thinking to rotate my certs and key every 5 days. I know device will reconnect using the new cert and key but what if someone uses the old cert and key and tries to provision the device. Will that also provision as logically the cert chain of the old cert is same the new one?

**Note**:- I don't have access to the windows or linux machine as that is of end customer so I don't have an option to save the cert and key file in TPM or HSM

**Comment by ggjjj** [comment]
@jlian Can you comment on the best practice for this scenario?

**Comment by jlian** [comment]
> Fetch a .pfx file from PKI and use that in edge config.toml. I think edge doesn't work with .pfx file and it needs the cert and private key to connect to IoT Hub. Also I didn't see any option in config.toml to mention the passphrase of the pfx file so that edge can decrpyt the pfx file itesef. Please confirm on this understanding is correct or not.

I believe you're correct

> I was thinking to have my custom encryption algorithm to encrypt the cert and private key and decrypt that during the installation process of edge before doing the provisioning. Once provisioning is done by edge device, I will delete the decrypted files. Question on this is - Is it that edge requires the certificate again anytime, as I think the provisioning is done only once and not again and again and cert and private key is meant for provisioning only.

Assuming you mean DPS, be careful about reprovisioning: https://learn.microsoft.com/en-us/azure/iot-edge/configure-device?view=iotedge-1.4#auto-reprovisioning-mode

https://learn.microsoft.com/en-us/azure/iot-dps/how-to-reprovision

> I was thinking to rotate my certs and key every 5 days. I know device will reconnect using the new cert and key but what if someone uses the old cert and key and tries to provision the device. Will that also provision as logically the cert chain of the old cert is same the new one?

The "old" cert/key is still unexpired (i.e. valid) and rooted in the CA that's tied to the enrollment group in this scenario, right? Then I believe yes, it will provision successfully.

> I don't have access to the windows or linux machine as that is of end customer so I don't have an option to save the cert and key file in TPM or HSM

That's too bad. That's our primary recommendation https://azure.microsoft.com/en-us/blog/hardening-the-security-of-azure-iot-edge-2/

**Comment by amit12cool** [comment]
@jlian Thanks for the answer. 
I also got to know that there is a option to encrypt the private key using open ssl and the encryption option of openssl is understood by IoT Hub and DPS. Is this type of feature also available on IoT Edge?

**Comment by jlian** [comment]
> I also got to know that there is a option to encrypt the private key using open ssl and the encryption option of openssl is understood by IoT Hub and DPS. Is this type of feature also available on IoT Edge?

Hmm, do you have a link?

**Comment by amit12cool** [comment]
@jlian https://github.com/Azure/azure-iot-sdk-c/blob/main/iothub_client/devdoc/iothubclient_c_library.md#openssl-engine-examples 
Above is the link for c sdk that I think does the job.

**Comment by jlian** [comment]
Oh ok yeah, the https://azure.github.io/iot-identity-service/ is part of IoT Edge, so it should be understood, yes

**Comment by amit12cool** [comment]
Please provide me an example, I don't think identity service is the right candidate for this.

Sent from Outlook for Android<https://aka.ms/AAb9ysg>
________________________________

**Comment by jlian** [comment]
Maybe this? https://azure.github.io/iot-identity-service/api/keys-service.html#encrypt

**Comment by jlian** [comment]
Actually @arsing can you help here? Probably should go through PKCS#11, right?

**Comment by jlian** [comment]
Quote from @arsing

> IoT Edge cannot use an openssl engine of their choice. The extension point for supporting arbitrary private keys is PKCS#11, not openssl engines.
> 
> They should figure out a TPM, HSM or something else to store their private key with the corresponding PKCS#11 library that can access / decrypt it, then follow PKCS#11 instructions.
> 
> If their reason for encrypting the key is just that they want encryption at rest, then they can just use standard Linux disk encryption (LUKS) instead of doing anything with openssl engines or PKCS#11.

**Comment by amit12cool** [comment]
@jlian can you point me to use the LUKS approach using the Azure iot c sdk.

**Comment by bishal41** [comment]
@arsing do you have a link that you can share with @amit12cool for the above ask?

**Comment by jlian** [comment]
@CIPop do you have any pointer on how to use standard Linux disk encryption with C SDK?

**Comment by buentead** [comment]
I tried to use the PKCS11 as documented in [iot-identity-service](https://azure.github.io/iot-identity-service/) also mentioned by @jlian . I wasn't successful due to library incompatibility with the OS I got. So I gave up.

My final solution was a small service that mounts an encrypted directory to `/var/lib/aziot` using gocryptfs. Make sure this service includes `Before=aziot-edged.service` so IoT Edge runtime waits for the mount. This solution is not production proofed yet. But runs since some days without any issue.

_Note:_ I don't know if this solution would also run on a Windows device. We are using Linux device only.

**Comment by jlian** [comment]
I got an email from @CIPop who is currently out of office with no access to GitHub

> The C-SDK doesn’t use any kind of storage, so disk encryption is out of scope. It could be used though by the application to load the very and keys from disk into memory.
> 
> Our recommendation is to store credentials within an HSM that has an OpenSSL Engine as driver.
> 
> Keys should be derived within the HSM and should never be downloaded (as the issue states). This is usually achieved using a CSR.
> 
> Examples of C-SDK OpenSSL Engine here: https://github.com/Azure/azure-iot-sdk-c/blob/main/iothub_client/devdoc/iothubclient_c_library.md#openssl-engine-examples

**Comment by jlian** [comment]
Closing, let us know if you need more and we can reopen
